qiskit/quantum_info/operators/symplectic/clifford_circuits.py

```diff
@@ -103,12 +103,24 @@ def _append_operation(clifford, operation, qargs=None):
         clifford.tableau = composed_clifford.tableau
         return clifford
 
-    # If not a Clifford basis gate we try to unroll the gate and
-    # raise an exception if unrolling reaches a non-Clifford gate.
-    # TODO: We could also check u3 params to see if they
-    # are a single qubit Clifford gate rather than raise an exception.
+    # If the gate is not in predefined basis gates but with up to 3 qubits,
+    # we try to construct a Clifford to be appended from its matrix representation.
+    if isinstance(gate, Gate) and len(qargs) <= 3:
```

Comment:
Hi, a quick question. With the current order of `if`-statements, would the new code be executed even for gates that already have a definition? I have not done any experiments, but it feels that constructing a Clifford from a matrix could be slower than constructing a Clifford from the defining quantum circuit.

Reply:
Good question. I know construction from `definition` is faster if possible but it does not always work because the `definition` may contain non-Clifford gates. For example, the definition of ECRGate contains `RZX(pi/4)`, which is not a Clifford. The `definition` of a gate is up to the gate designer, so we cannot fully rely on it. And I just prioritize capability to performance in this commit.

Reply:
Another suggestion would be to manually define all the relevant gates here:
https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/symplectic/clifford_circuits.py
(you can check both gate name and parameters, and see if they fit one of the known patterns)

Reply:
Thanks! I was merely suggesting to change the order of checks:

```
if gate.definition is not None:
    return _append_circuit(clifford, gate.definition, qargs)

try:
   # new code here

raise QiskitError(f"Cannot apply {gate}")
```

Reply:
If we change the order of check, we may need to copy `clifford` and when we can any exception, i.e. we fails to all-Clifford decomposition (based on `definition`), we need to restore the clifford that was before trying unroll. So the actual code would require more lines.

Reply:
@alexanderivrii I've tried the implementation that changes the order of check in [this branch](https://github.com/itoko/qiskit-terra/pull/49). I've confirmed that it gets faster instead of increasing the complexity of the implementation. Do you want to include the change in this PR?

Reply:
It's generally good to add more gate (like `Rz(pi/2)` and `u` for angles that are multiples of `pi/2`) to `clifford_circuits.py`.
My PR #9623 may also be useful. Would you like to add other gates (also with parameters) directly to `clifford_circuits.py`?

Reply:
@ShellyGarion I'm not thinking adding more `_append_*` to `clifford_circuits.py` is the best approach but it should a option to be considered for improving the performance of `Clifford.from_circuit` (see [my comment in #9582](https://github.com/Qiskit/qiskit-terra/issues/9582#issuecomment-1438330259) for other options). Let's continue to discuss it in #9582.

Reply:
Regarding the order of check, I've added at [b64d969](https://github.com/Qiskit/qiskit-terra/pull/9475/commits/b64d969a7c9da839f644fa162848134557dcf5f1) with the conversion of U gate [c4bc00a](https://github.com/Qiskit/qiskit-terra/pull/9475/commits/c4bc00a4c8c78ea468b3715d900c25bf4176b41d)